docs/proposal/2022-09-28-event-manager-extensibility.md

```diff
@@ -0,0 +1,302 @@
+process of that plugin. They might not be a simple Go program, that's why there is no unified `Makefile` that builds any Go project, it can contain
+customer specific build flows. For `kubectl` and `Kubernetes`, they are same for now. 
+
+### How to build plugins?
+`plugins` folder only contains the source code of the plugins, and consumer (Botkube or playground project on our case), needs the executable versions
+of those plugin to call via Hashicorp's plugin package. There is a Github Action for plugin release flow which you can see [here](https://github.com/huseyinbabal/Botkube-plugins/blob/main/.github/workflows/release.yaml). 
+In this flow, the only thing you need to do is pushing a new tag in a format `<plugin_name>/<version>` like `Kubernetes/v1.0.7`. This will build `Kubernetes` plugin, create a release with version
+`v1.0.7` and go binary executable will be attached to release artifacts. Notice that, those artifacts are used in consumer side to be used as parameter to Hashicorp's Go Plugin.
+
+### How can I know the metadata of plugins?
+In `Botkube-plugins` project, we have a plugin index file as you can also see it [here](https://github.com/huseyinbabal/Botkube-plugins/blob/main/index.json). That contains basic metadata of each plugin, and notice that they are managed manually
+right now, we can automate and do find/replace whenever there is a change to plugins. Here is an example;
+```json
+[
+  {
+    "name": "Kubernetes",
+    "type": "source",
+    "description": "Kubernetes source plugin to consume events from Kubernetes Events API.",
+    "version": "v1.0.9"
+  },
+  {
+    "name": "kubectl",
+    "type": "executor",
+    "description": "Kubectl executor plugin to handle given command in configured k8s cluster.",
+    "version": "v1.0.9"
+  }
+]
+```
+
+### How can I provide plugin configuration?
+If you use terraform before, they use a notation like `<pluginname>_...` to pass settings to specific plugin. So, in Botkube, if somebody provides `KUBERNETES_CONFIG_PATH=something`, the `kubernetes` plugin executable can be run as follows.
+```go
+client := plugin.NewClient(&plugin.ClientConfig{
+    Plugins:          m.pluginMap(metadata),
+    VersionedPlugins: nil,
+    Cmd:              exec.Command(metadata.Path, "--config", "something"), // Plugin specific params goes here
+    AllowedProtocols: []plugin.Protocol{plugin.ProtocolGRPC},
+    HandshakeConfig: plugin.HandshakeConfig{
+        ProtocolVersion:  1,
+        MagicCookieKey:   "BOTKUBE_MAGIC_COOKIE",
+        MagicCookieValue: "BOTKUBE_BASIC_PLUGIN",
+    },
+})
+```
+
+`type` can be `source` or `executor` since we decide on plugin implementation contract based on that value.
+
+### How to install/uninstall plugins?
+Once you initialize plugin, it collects plugin metadatas from plugin index and initializes all of them by using Hashicorp's plugin system.
```

Comment:
So in the first implementation, each change will simply restart the Botkube as we do right now?

Reply:
Not necessarily, there is an api in Hashicorp's Go plugin, I haven't fully tested that, the process would be update configuration and call Hashicorp plugin API to remove it.